### PR TITLE
oled improvement revival

### PR DESCRIPTION
**Pull request by after-ephemera**
This is meant to be strict revival of https://github.com/KMKfw/kmk_firmware/pull/548 so we can land these improvements. I do not have a kyria but [I expect to use the powersave disable feature on my lily58](https://kmkfw.zulipchat.com/#narrow/stream/358347-general/topic/oled.20powersave.20mode). 

If any features are still up for debate per the previous PR, I'm more than happy to try to pare it down to just the powersaving ones that I need. Thanks to @janlindblom for doing the heavy lifting here.

### CODE REVIEW

File: kmk/extensions/oled.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+    def on_runtime_enable(self, sandbox):
+        return
+
+    def on_runtime_disable(self, sandbox):
+        return
+
+    def during_bootup(self, board):
+        displayio.release_displays()
+        i2c = busio.I2C(board.SCL, board.SDA)
```
Comment (regicidalplutophage) [inline kmk/extensions/oled.py:119]:
```suggestion
        i2c = busio.I2C(keyboard.SCL, keyboard.SDA)
```
We define all the pins in keyboard class anyway, this way we can use I2C pins that aren't default.

File: boards/kyria/main.py [hunk 2]
```diff
@@ -1,42 +1,64 @@
...
+            OledData.oled_text_entry(text='Kyria v1.4', x=0, y=0),
+            OledData.oled_text_entry(text='KB2040', x=0, y=10),
+            OledData.oled_text_entry(text='Layer: ', x=0, y=20),
+            OledData.oled_text_entry(text='BASE', x=40, y=20, layer=0),
+            OledData.oled_text_entry(text='LOWER', x=40, y=20, layer=3),
+            OledData.oled_text_entry(text='RAISE', x=40, y=20, layer=4),
+            OledData.oled_text_entry(text='ADJUST', x=40, y=20, layer=6),
+        ]
+    ),
+    oHeight=64,
```
Comment (regicidalplutophage) [inline boards/kyria/main.py:48]:
```suggestion
    height=64,
```

File: kmk/extensions/oled.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+        self.render_oled(sandbox.active_layers[0])
+        gc.collect()
+
+    def on_runtime_enable(self, sandbox):
+        return
+
+    def on_runtime_disable(self, sandbox):
+        return
+
+    def during_bootup(self, board):
```
Comment (regicidalplutophage) [inline kmk/extensions/oled.py:117]:
```suggestion
    def during_bootup(self, keyboard):
```
Forgot that!

**Comment by regicidalplutophage** [comment]
Love that this is being revived. Toying with this on my keyboard and so far I'm not sure I get the powersave feature.
It dims the display if powersave is enabled and that's it. Ideally I'd want it to shut the display off if a keyboard hasn't been touched in some time.
Otherwise, I quite like everything else.

**Comment by regicidalplutophage** [comment]
I made a PR to your fork, it doesn't address the behavior on powersave enable, but adds a bunch of features to text entries.

**Comment by after-ephemera** [comment]
I'm going to mess with the powersave again and see if I can get it working consistently on my board and then come back around here to see if we can't get this through.

**Comment by regicidalplutophage** [comment]
Honestly I think it can be merged for what it is, but I'd also like to try some stuff. I experimented a bit trying to make it work it portrait mode, couldn't manage that so far. I'd also like to try and and make a progressive dimming function independent of powersave.

**Comment by regicidalplutophage** [comment]
I have successfully implemented timed dimming function, then discovered that reducing brightness to 0 doesn't black out the screen, Will have to look further some time soon. For now, making a RP...

**Pull request closed without merging**